File: backend/travel_platform/revenue/dynamic_pricing.py
"""
Dynamic pricing engine — adjusts unit price from seat occupancy vs trip capacity.
"""

from __future__ import annotations

from dataclasses import dataclass
from decimal import Decimal, ROUND_HALF_UP
from enum import Enum
from uuid import UUID

from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from core.base_service import TenantScopedService
from core.config import platform_settings
from core.exceptions import PricingError


class PricingRuleType(str, Enum):
    HIGH_OCCUPANCY_SURGE = "high_occupancy_surge"
    LOW_OCCUPANCY_DISCOUNT = "low_occupancy_discount"
    FLAT = "flat"


@dataclass(frozen=True)
class TripCapacitySnapshot:
    trip_id: int
    total_seats: int
    sold_seats: int
    base_price_eur: Decimal

    @property
    def occupancy_ratio(self) -> float:
        if self.total_seats <= 0:
            return 0.0
        return min(1.0, self.sold_seats / self.total_seats)


@dataclass(frozen=True)
class PricingQuote:
    trip_id: int
    base_price_eur: Decimal
    final_price_eur: Decimal
    occupancy_ratio: float
    applied_rule: PricingRuleType | None
    adjustment_pct: float


def compute_quote_pure(
    base_price_eur: float,
    total_seats: int,
    sold_seats: int,
    *,
    trip_id: int = 0,
    high_threshold: float | None = None,
    high_markup_pct: float | None = None,
    low_threshold: float | None = None,
    low_discount_pct: float | None = None,
) -> PricingQuote:
    """Pure pricing — no DB (B2C / admin quote endpoint)."""
    high_threshold = (
        high_threshold if high_threshold is not None else platform_settings.pricing_high_occupancy_threshold
    )
    high_markup = (
        high_markup_pct if high_markup_pct is not None else platform_settings.pricing_high_occupancy_markup_pct
    )
    low_threshold = (
        low_threshold if low_threshold is not None else platform_settings.pricing_low_occupancy_threshold
    )
    low_discount = (
        low_discount_pct if low_discount_pct is not None else platform_settings.pricing_low_occupancy_discount_pct
    )
    capacity = max(total_seats, 1)
    ratio = min(1.0, sold_seats / capacity)
    base = Decimal(str(base_price_eur))
    rule: PricingRuleType | None = None
    adjustment_pct = 0.0
    if ratio >= high_threshold:
        rule = PricingRuleType.HIGH_OCCUPANCY_SURGE
        adjustment_pct = high_markup
    elif ratio <= low_threshold:
        rule = PricingRuleType.LOW_OCCUPANCY_DISCOUNT
        adjustment_pct = -low_discount
    multiplier = Decimal("1") + Decimal(str(adjustment_pct)) / Decimal("100")
    final = (base * multiplier).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
    return PricingQuote(
        trip_id=trip_id,
        base_price_eur=base,
        final_price_eur=final,
        occupancy_ratio=ratio,
        applied_rule=rule,
        adjustment_pct=adjustment_pct,
    )


class DynamicPricingEngine(TenantScopedService):
    """Stateless strategy evaluation; persists quotes optionally via audit."""

    def __init__(self, session: AsyncSession, tenant_id: UUID, **kwargs):
        super().__init__(session, tenant_id, **kwargs)
        self._high_threshold = platform_settings.pricing_high_occupancy_threshold
        self._high_markup = platform_settings.pricing_high_occupancy_markup_pct
        self._low_threshold = platform_settings.pricing_low_occupancy_threshold
        self._low_discount = platform_settings.pricing_low_occupancy_discount_pct

    async def load_trip_snapshot(self, trip_id: int) -> TripCapacitySnapshot:
        await self._bind_tenant_rls()
        row = await self._session.execute(
            text("""
                SELECT t.id AS trip_id,
                       t.total_seats,
                       COUNT(b.id) FILTER (WHERE b.status IN ('CONFIRMED','PENDING')) AS sold_seats,
                       t.base_price AS base_price_eur
                FROM trips t
                LEFT JOIN bookings b ON b.trip_id = t.id AND b.tenant_id = t.tenant_id
                WHERE t.id = :trip_id AND t.tenant_id = :tenant_id
                GROUP BY t.id, t.total_seats, t.base_price
            """),
            {"trip_id": trip_id, "tenant_id": str(self._tenant_id)},
        )
        mapping = row.mappings().first()
        if not mapping:
            raise PricingError(f"Trip {trip_id} not found for tenant")
        return TripCapacitySnapshot(
            trip_id=mapping["trip_id"],
            total_seats=mapping["total_seats"] or 0,
            sold_seats=mapping["sold_seats"] or 0,
            base_price_eur=Decimal(str(mapping["base_price_eur"])),
        )

    def compute_quote(self, snapshot: TripCapacitySnapshot) -> PricingQuote:
        return compute_quote_pure(
            float(snapshot.base_price_eur),
            snapshot.total_seats,
            snapshot.sold_seats,
            trip_id=snapshot.trip_id,
            high_threshold=self._high_threshold,
            high_markup_pct=self._high_markup,
            low_threshold=self._low_threshold,
            low_discount_pct=self._low_discount,
        )

    async def quote_for_trip(self, trip_id: int) -> PricingQuote:
        snapshot = await self.load_trip_snapshot(trip_id)
        quote = self.compute_quote(snapshot)
        await self._audit(
            "pricing.quote_generated",
            "trip",
            str(trip_id),
            metadata={
                "occupancy_ratio": quote.occupancy_ratio,
                "final_price_eur": str(quote.final_price_eur),
                "rule": quote.applied_rule.value if quote.applied_rule else None,
            },
            financial=True,
        )
        return quote

File: src/lib/revenue/dynamicPricing.js
import { DEFAULT_PLATFORM_SETTINGS } from '../../services/platformApi.js';
import { loadBookings, isBookingPaid } from '../ticketing/bookingStore.js';
import { generateSeatMap } from '../seats/generateSeatMap.js';

/**
 * @param {import('../../data/mockData.js').mockTrips[0]} trip
 */
export function getTripSeatCapacity(trip) {
  if (!trip) return 45;
  const { seats } = generateSeatMap(trip);
  return seats.length || trip.availableSeats || 45;
}

/**
 * @param {import('../../data/mockData.js').mockTrips[0]} trip
 */
export function countSoldSeatsForTrip(trip) {
  if (!trip) return 0;
  const { seats } = generateSeatMap(trip);
  const fromMap = seats.filter((s) => s.status === 'BOOKED').length;

  const fromBookings = loadBookings()
    .filter(
      (b) =>
        isBookingPaid(b) &&
        (b.tripId === trip.id || b.tripTitle === trip.title),
    )
    .reduce((sum, b) => {
      if (Array.isArray(b.seats) && b.seats.length) return sum + b.seats.length;
      if (b.seat) return sum + b.seat.split(',').filter(Boolean).length;
      return sum + 1;
    }, 0);

  return Math.max(fromMap, fromBookings);
}

/**
 * @param {import('../../data/mockData.js').mockTrips[0]} trip
 * @param {typeof DEFAULT_PLATFORM_SETTINGS} settings
 */
export function computeDynamicPrice(trip, settings = DEFAULT_PLATFORM_SETTINGS) {
  const basePrice = Number(trip?.price) || 0;
  const totalSeats = getTripSeatCapacity(trip);
  const soldSeats = countSoldSeatsForTrip(trip);
  const capacity = Math.max(totalSeats, 1);
  const occupancyRatio = Math.min(1, soldSeats / capacity);

  const highThreshold = Number(settings.pricing_high_occupancy_threshold ?? 0.8);
  const highMarkup = Number(settings.pricing_high_occupancy_markup_pct ?? 10);
  const lowThreshold = Number(settings.pricing_low_occupancy_threshold ?? 0.3);
  const lowDiscount = Number(settings.pricing_low_occupancy_discount_pct ?? 5);

  let adjustmentPct = 0;
  let appliedRule = null;

  if (occupancyRatio >= highThreshold) {
    appliedRule = 'high_occupancy_surge';
    adjustmentPct = highMarkup;
  } else if (occupancyRatio <= lowThreshold) {
    appliedRule = 'low_occupancy_discount';
    adjustmentPct = -lowDiscount;
  }

  const finalPrice =
    Math.round(basePrice * (1 + adjustmentPct / 100) * 100) / 100;

  return {
    tripId: trip?.id,
    basePrice,
    finalPrice,
    occupancyRatio,
    adjustmentPct,
    appliedRule,
    soldSeats,
    totalSeats,
    hasAdjustment: adjustmentPct !== 0,
  };
}

export function pricingRuleLabel(appliedRule) {
  if (appliedRule === 'high_occupancy_surge') return 'Υψηλή ζήτηση';
  if (appliedRule === 'low_occupancy_discount') return 'Προσφορά';
  return null;
}
